lib/strategy/FragmentationStrategyShape.ts

```diff
@@ -0,0 +1,292 @@
+    const shapes = config.shapes;
+    for (const [ dataType, shapeEntry ] of Object.entries(shapes)) {
+      shapeMap.set(dataType, {
+        shape: readFileSync(join(shapeFolder, (<IShapeEntry>shapeEntry).shape)).toString(),
+        directory: (<IShapeEntry>shapeEntry).directory,
+      });
+    }
+    return shapeMap;
+  }
+
+  /**
+   * Determine if the shape information should be generated.
+   * returns always true if the probability is not defined.
+   * @returns {boolean} indicate if the shape information should be generated
+   */
+  public shouldGenerate(): boolean {
+    return this.generationProbability === undefined ?
+      true :
+      Math.random() * 100 <= this.generationProbability;
+  }
+
+  /**
+   * From a quad generate the IRI based on the {@link FragmentationStrategySubject}.
+   * Evaluate if shape index information should be generated based on which quad has been handled and
+   * the shape config file provides. Generate the right iri for the shape indexes.
+   * The generation of the triples and the quad sink pushing is handled by other functions called by
+   * handleQuad.
+   * @param {RDF.Quad} quad - a quad
+   * @param {IQuadSink} quadSink - a quad sink
+   */
+  protected async handleQuad(quad: RDF.Quad, quadSink: IQuadSink): Promise<void> {
+    const iri = FragmentationStrategySubject.generateIri(quad, this.relativePath);
+    // If iri already has been handled, we do nothing
+    if (!this.irisHandled.has(iri)) {
+      for (const [ resourceIndex, { shape, directory }] of this.shapeMap) {
+        // Find the position of the first character of the container
+        const positionContainerResourceNotInRoot = iri.indexOf(`/${directory}/`);
+        const positionContainerResourceInRoot = iri.indexOf(`/${resourceIndex}`);
+        const positionContainer = positionContainerResourceNotInRoot !== -1 ?
+          positionContainerResourceNotInRoot :
+          positionContainerResourceInRoot;
+
+        // We use as an id the path of the iri until the resource identifier.
+        // It is a different discrimination mechanism than the one related to the iri because even
+        // if we don't want to add shape index information we might want to add shapeTreeLocator in
+        // every resource bounded by a shape.
+        const resourceId = `${iri.slice(0, Math.max(0, positionContainer))}/${resourceIndex}`;
+        if (positionContainer !== -1) {
+          const generate = this.shouldGenerate();
+          // Skip the resource based on the generation probability
+          if (!generate) {
+            this.irisHandled.add(iri);
+            this.resourcesHandled.add(resourceId);
+            return;
+          }
+          const podIRI = iri.slice(0, Math.max(0, positionContainer));
+
+          const shapeTreeIRI = `${podIRI}/${FragmentationStrategyShape.shapeTreeFileName}`;
+
+          const promises: Promise<void>[] = [];
```

Comment:
Not sure I see the value of this promise array. Since it will only contain at most 2 values. Can't we just await each promise directly inline?